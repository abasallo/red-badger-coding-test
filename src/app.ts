import * as dotenv from 'dotenv'

import express from 'express'
import cors from 'cors'
import bodyParser from 'body-parser'

import index from './express/routes'
import robots from './express/routes/robots'

import { errorMiddleware } from './express/errors'

dotenv.config()

export const app = express()

app.use(cors())

app.use(bodyParser.urlencoded({ extended: false }))
app.use(bodyParser.json())

app.use('/', index)
app.use('/robots', robots)

app.use(errorMiddleware)

export default app
