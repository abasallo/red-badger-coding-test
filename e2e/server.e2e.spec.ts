import axios from 'axios'
import http from 'http'
import { ReasonPhrases, StatusCodes } from 'http-status-codes'

import app from '../src/app'

const TEST_SERVER_PORT = 4001

const testData = {
  upperRight: { x: 5, y: 3 },
  commands: [
    {
      origin: {
        coordinate: { x: 1, y: 1 },
        orientation: 'E'
      },
      instructions: ['R', 'F', 'R', 'F', 'R', 'F', 'R', 'F']
    },
    {
      origin: {
        coordinate: { x: 3, y: 2 },
        orientation: 'N'
      },
      instructions: ['F', 'R', 'R', 'F', 'L', 'L', 'F', 'F', 'R', 'R', 'F', 'L', 'L']
    },
    {
      origin: {
        coordinate: { x: 0, y: 3 },
        orientation: 'W'
      },
      instructions: ['L', 'L', 'F', 'F', 'F', 'L', 'F', 'L', 'F', 'L']
    }
  ]
}

let server: http.Server
beforeAll(async () => {
  server = app.listen({ port: TEST_SERVER_PORT }, () => console.log(`Test Server initialised on port: ${TEST_SERVER_PORT}`))
})

afterAll(() => {
  server.close()
})

describe('Server E2E tests', () => {
  describe('Health Check E2E tests', () => {
    test('Expected response for Health Check', async () => {
      const { status, statusText, data } = await axios({
        method: 'get',
        url: `http://localhost:${TEST_SERVER_PORT}/`
      })

      expect(status).toEqual(StatusCodes.OK)
      expect(statusText).toEqual(ReasonPhrases.OK)

      expect(data).toMatchSnapshot()
    })
  })

  describe('Robots E2E tests', () => {
    test('Expected results for provided sample data', async () => {
      const { status, statusText, data } = await axios({
        method: 'get',
        url: `http://localhost:${TEST_SERVER_PORT}/robots`,
        data: testData
      })

      expect(status).toEqual(StatusCodes.OK)
      expect(statusText).toEqual(ReasonPhrases.OK)

      expect(data).toMatchSnapshot()
    })

    test('Wrong format payload', async () => {
      try {
        await axios({
          method: 'get',
          url: `http://localhost:${TEST_SERVER_PORT}/robots`,
          data: { id: 'wrong' }
        })
      } catch (error) {
        expect(error).toBeDefined()
      }
    })
  })
})
