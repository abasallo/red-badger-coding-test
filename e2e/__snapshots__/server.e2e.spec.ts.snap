// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Server E2E tests Health Check E2E tests Expected response for Health Check 1`] = `"OK"`;

exports[`Server E2E tests Robots E2E tests Expected results for provided sample data 1`] = `
[
  "1 1 E",
  "3 3 N LOST",
  "2 3 S",
]
`;
